File: Loenn/entities/cassetteIntroCar.lua
local drawableSpriteStruct = require("structs.drawable_sprite")
local colors = require("consts.xna_colors")
local utils = require("utils")

local introCar = {}

local barrierTexture = "scenery/car/barrier"
local bodyTexture = "scenery/car/body"
local pavementTexture = "scenery/car/pavement"
local wheelsTexture = "scenery/car/wheels"

local colors = {
    {73 / 255, 170 / 255, 240 / 255},
    {240 / 255, 73 / 255, 190 / 255},
    {252 / 255, 220 / 255, 58 / 255},
    {56 / 255, 224 / 255, 78 / 255},
}

local depth = 1

local colorNames = {
    ["Blue"] = 0,
    ["Rose"] = 1,
    ["Bright Sun"] = 2,
    ["Malachite"] = 3
}

introCar.name = "brokemiahelper/cassetteIntroCar"

introCar.fieldInformation = {
    index = {
        fieldType = "integer",
        options = colorNames,
        editable = false
    }
}
introCar.placements = {}

for i, _ in ipairs(colors) do
    introCar.placements[i] = {
        name = string.format("cassette_intro_car_%s", i - 1),
        data = {
            index = i - 1,
            tempo = 1.0,
            hasRoadAndBarriers = false,
            width = 16,
            height = 16
        }
    }
end

function introCar.sprite(room, entity)
    local sprites = {}
    local x, y = entity.x, entity.y
    local hasRoadAndBarriers = entity.hasRoadAndBarriers
    local index = entity.index or 0
    local color = colors[index + 1] or colors[1]

    local bodySprite = drawableSpriteStruct.fromTexture(bodyTexture, entity)
    bodySprite:setJustification(0.5, 1.0)
    bodySprite:setColor(color)
    bodySprite.depth = depth

    local wheelSprite = drawableSpriteStruct.fromTexture(wheelsTexture, entity)
    wheelSprite:setJustification(0.5, 1.0)
    bodySprite:setColor(color)
    wheelSprite.depth = depth + 2

    table.insert(sprites, bodySprite)
    table.insert(sprites, wheelSprite)

    if hasRoadAndBarriers then
        utils.setSimpleCoordinateSeed(x, y)

        local pavementWidth = x - 48
        local columns = math.floor(pavementWidth / 8)

        for column = 0, columns - 1 do
            local choice = column >= columns - 2 and (column ~= columns - 2 and 3 or 2) or math.random(0, 2)
            local pavementSprite = drawableSpriteStruct.fromTexture(pavementTexture, entity)
            pavementSprite:addPosition(column * 8 - x, 0)
            pavementSprite.depth = -10001
            pavementSprite:setJustification(0.0, 0.0)
            pavementSprite:useRelativeQuad(choice * 8, 0, 8, 8)

            table.insert(sprites, pavementSprite)
        end

        local barrier1Sprite = drawableSpriteStruct.fromTexture(barrierTexture, entity)
        barrier1Sprite:addPosition(32, 0)
        barrier1Sprite:setJustification(0.0, 1.0)
        barrier1Sprite.depth = -10

        local barrier2Sprite = drawableSpriteStruct.fromTexture(barrierTexture, entity)
        barrier2Sprite:addPosition(41, 0)
        barrier2Sprite:setJustification(0.0, 1.0)
        barrier2Sprite.depth = 5
        barrier2Sprite.color = colors.DarkGray

        table.insert(sprites, barrier1Sprite)
        table.insert(sprites, barrier2Sprite)
    end

    return sprites
end

function introCar.selection(room, entity)
    local bodySprite = drawableSpriteStruct.fromTexture(bodyTexture, entity)
    local wheelSprite = drawableSpriteStruct.fromTexture(wheelsTexture, entity)

    bodySprite:setJustification(0.5, 1.0)
    wheelSprite:setJustification(0.5, 1.0)

    return utils.rectangle(utils.coverRectangles({bodySprite:getRectangle(), wheelSprite:getRectangle()}))
end

return introCar

File: Ahorn/entities/cassetteIntroCar.jl
module CassetteIntroCar

using ..Ahorn, Maple

@mapdef Entity "brokemiahelper/cassetteIntroCar" CarCassette(x::Integer, y::Integer, index::Integer=0, tempo::Number=1)

colorNames = Dict{String, Int}(
    "Blue" => 0,
    "Rose" => 1,
    "Bright Sun" => 2,
    "Malachite" => 3
)

const placements = Ahorn.PlacementDict(
    "Cassette Intro Car ($index - $color, BrokemiaHelper)" => Ahorn.EntityPlacement(
        CarCassette,
        "point",
        Dict{String, Any}(
            "index" => index,
        )
    ) for (color, index) in colorNames
)

Ahorn.editingOptions(entity::CarCassette) = Dict{String, Any}(
    "index" => colorNames
)

colors = Dict{Integer, Ahorn.colorTupleType}(
    1 => (240, 73, 190, 255) ./ 255,
	2 => (252, 220, 58, 255) ./ 255,
	3 => (56, 224, 78, 255) ./ 255,
)

defaultColor = (73, 170, 240, 255) ./ 255
borderMultiplier = (0.9, 0.9, 0.9, 1)

barrierSprite = "scenery/car/barrier"
bodySprite = "scenery/car/body"
pavementSprite = "scenery/car/pavement"
wheelsSprite = "scenery/car/wheels"

function Ahorn.selection(entity::CarCassette)
    x, y = Ahorn.position(entity)

    hasRoadAndBarriers = get(entity.data, "hasRoadAndBarriers", false)

    rectangles = Ahorn.Rectangle[
        Ahorn.getSpriteRectangle(bodySprite, x, y, jx=0.5, jy=1.0),
        Ahorn.getSpriteRectangle(wheelsSprite, x, y, jx=0.5, jy=1.0),
    ]

    if hasRoadAndBarriers
        push!(rectangles, Ahorn.getSpriteRectangle(barrierSprite, x + 32, y, jx=0.0, jy=1.0))
        push!(rectangles, Ahorn.getSpriteRectangle(barrierSprite, x + 41, y, jx=0.0, jy=1.0))
    end

    return Ahorn.coverRectangles(rectangles)
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::CarCassette, room::Maple.Room)
    x, y = Ahorn.position(entity)

    hasRoadAndBarriers = get(entity.data, "hasRoadAndBarriers", false)
    rng = Ahorn.getSimpleEntityRng(entity)

    pavementWidth = x - 48
    columns = floor(Int, pavementWidth / 8)

    Ahorn.drawSprite(ctx, wheelsSprite, x, y, jx=0.5, jy=1.0, tint=get(colors, Int(get(entity.data, "index", 0)), defaultColor))
    Ahorn.drawSprite(ctx, bodySprite, x, y, jx=0.5, jy=1.0, tint=get(colors, Int(get(entity.data, "index", 0)), defaultColor))

    if hasRoadAndBarriers
        Ahorn.drawSprite(ctx, barrierSprite, x + 32, y, jx=0.0, jy=1.0, tint=get(colors, Int(get(entity.data, "index", 0)), defaultColor))
        Ahorn.drawSprite(ctx, barrierSprite, x + 41, y, jx=0.0, jy=1.0, tint=get(colors, Int(get(entity.data, "index", 0)), defaultColor))

        for col in 0:columns - 1
            choice = col >= columns - 2 ? (col != columns - 2 ? 3 : 2) : rand(rng, 0:2)

            Ahorn.drawImage(ctx, pavementSprite, col * 8, y, choice * 8, 0, 8, 4, tint=get(colors, Int(get(entity.data, "index", 0)), defaultColor))
        end
    end
end

end
